### PR TITLE
Fix unclear documentation and indicate variable-length argument lists.

### PR DESCRIPTION
**Pull request by frederikwilde**
**Context:**
The documentation about [adding operators](https://github.com/PennyLaneAI/pennylane/blob/master/doc/development/adding_operators.rst?plain=1#L98) and [`Operator.compute_matrix()`](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/operation.py#L708) contain some confusing notes.
Variable-length argument lists (e.g. `foo(*args)`) are not always indicated in the respective docstrings.

**Description of the Change:**
The documentation is improved. Variable-length argument lists are indicated according to the [Google Styleguide](https://google.github.io/styleguide/pyguide.html).

**Benefits:**
Users can understand more easily how to pass arguments to functions and methods. Confusing notes are resolved.

**Possible Drawbacks:**
n/a

**Related GitHub Issues:**
n/a

### CODE REVIEW

**Comment by frederikwilde** [comment]
[sc-38671]

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -342,6 +342,10 @@
   An if conditional was intended to prevent divide by zero errors but the division was by the sine of the argument so any multiple of $\pi$ should trigger the conditional, but it was only checking if the argument was 0. Example: `qml.Rot(2.3, 2.3, 2.3)`
   [(#4210)](https://github.com/PennyLaneAI/pennylane/pull/4210)
 
+* Fix unclear documentation and indicate variable-length argument lists of functions and methods in
+  the respective docstrings.
+  [(#4242)](https://github.com/PennyLaneAI/pennylane/pull/4242)
+
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:348]: Maybe this should be in improvements?

File: pennylane/operation.py [hunk 2]
```diff
@@ -762,8 +762,8 @@ def compute_sparse_matrix(*params, **hyperparams):  # pylint:disable=unused-argu
         .. seealso:: :meth:`~.Operator.sparse_matrix`
 
         Args:
-            params (list): trainable parameters of the operator, as stored in the ``parameters`` attribute
-            hyperparams (dict): non-trainable hyperparameters of the operator, as stored in the ``hyperparameters``
+            *params (list): trainable parameters of the operator, as stored in the ``parameters`` attribute
+            **hyperparams (dict): non-trainable hyperparameters of the operator, as stored in the ``hyperparameters``
```
Comment (Jaybsoni) [inline pennylane/operation.py:766]: Actually is this the standard way to denote variable length arguments? This is the first time I have ever seen it denoted this way in the docs? Might be worth asking @albi3ro if this is the standard approach we want to take from now on?

**Review by Jaybsoni — COMMENTED** [review]
Looks good, just two small comments, happy to approve once addressed 👍🏼

**Review by lillian542 — APPROVED** [review]
Looks good to me! 🎉

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (frederikwilde) [reply doc/releases/changelog-dev.md:348]: Oh, yes. I moved it.

File: pennylane/operation.py (hunk 2 above)
Comment (frederikwilde) [reply pennylane/operation.py:766]: I've never actually looked for an official standard, but the format chosen in PennyLane matches the one introduced by Google (see [this post](https://stackoverflow.com/questions/3898572/what-are-the-most-common-python-docstring-formats) for an overview of formats).
According to the [Google style guide](https://google.github.io/styleguide/pyguide.html) variable-length arguments and generic keyword arguments should be denoted as `*args` and `**kwargs`, respectively.
One could also choose a different convention, but some form of distinction from fixed-length arguments is more user friendly, I'd say.

Comment (albi3ro) [reply pennylane/operation.py:766]: Screenshot 2023-06-13 at 3 50 58 PM

Looks like the style guide says we should be doing it this way. I actually think it much looks much clearer this way. There's probably a lot of places in pennylane

Comment (albi3ro) [reply pennylane/operation.py:766]: I actually really like this way of denoting variable length arguments and keyword arguments. We probably have a bunch of places in our documentation that need to be altered.

We should probably start by letting more people know this is the way we should actually do things.

**Review by Jaybsoni — APPROVED** [review]
💯

File: pennylane/operation.py (hunk 2 above)
Comment (frederikwilde) [reply pennylane/operation.py:766]: I regex-matched functions that have variable-length argument lists across PennyLane and replaced them. But I probably missed some.

**Pull request merged**